### DISCUSSION TITLE
[xtensa] support multiple config using the same xt-sim board

### DISCUSSION THREAD

**Discussion by williamtcdns** [zephyrproject-rtos/zephyr#78198] [Q&A]
When using Xtensa tools the environment variable `XTENSA_CORE` must be set to the config to use.

Currently the board `xt-sim`  is hardwired to use `sample_controller`, and `XTENSA_CORE` must be set to `sample_controller`.

I would like to support the config `hifi5s_ao_7` using the board `xt-sim`.
I implemented the support and successfully built and ran `samples/hello_world` in the following commit:
https://github.com/williamtcdns/zephyr/commit/db88e999ed6
However, in doing so, I have to replace the hardwired use of `sample_controller` by `hifi5s_ao_7` by editing 3 files:
- boards/cdns/xt-sim/Kconfig.xt-sim
- boards/cdns/xt-sim/board.yml
- boards/cdns/xt-sim/xt-sim.dts

Is it possible to have above 3 edited files automatically select between `sample_controller` and `hifi5s_ao_7` based on the value of the environment variable `XTENSA_CORE` ?

**Comment by williamtcdns** [comment]
@henrikbrixandersen @dcpleung @andyross @nashif @ceolin

**Comment by williamtcdns** [comment]
Please note that config `hifi5s_ao_7` is available in Xtensa tools RJ-2024.3 .

**Comment by dcpleung** [comment]
xt-sim is a simulator and is not a board. If you need to support a new SoC, you need to create a new SoC configuration and corresponding board configuration. In the board configuration, you can specify that you want to use xt-sim as the simulator.

**Reply by williamtcdns** [reply]
Yes `xt-sim` is a simulator, but there is a board by the same name implemented at `boards/cdns/xt-sim` such that it can be used as follow `west build -b xt-sim samples/hello_world` .

In the commit https://github.com/williamtcdns/zephyr/commit/db88e999ed6 , I did create a new SoC configuration under `soc/cdns/hifi5s_ao_7` and specified it in the `xt-sim` board configuration, but in doing so, I have to replace the `xtensa_sample_controller` SoC currently specified in the `xt-sim` board configuration.

My request for help is whether it is possible to have a board support multiple SoC.

**Reply by williamtcdns** [reply]
It seems to be possible when I look at `boards/digilent/arty_a7/Kconfig.arty_a7` which selects between 2 SoCs.

In my case, for `boards/cdns/xt-sim`, I would like to select between the SoCs `xtensa_sample_controller` and `hifi5s_ao_7` based on the value of the environment variable `XTENSA_CORE`.

If possible, could you provide an outline of the files to use or changes to make ?

**Comment by dcpleung** [comment]
Maybe you can follow the arty_a7 board and see what changes you can make to enable your usage?

**Comment by andyross** [comment]
We don't really have a clean answer here.  The metaphor is that boards/soc/arch is fixed hierarchically.  Strictly you'd want a separate board variant defined for every XTENSA_CORE, which is exactly what you don't want.

This is unlikely to work without surgery.  Note that the DTS is downstream of kconfig, so it would be possible to define e.g. a "ARCH_XTENSA_CORE_ID" variable which you'd then select on in the dts, which can reduce some boilerplate. Then if you wanted to use the $XTENSA_CORE variable in the environment, you'd need to extract it in the cmake somewhere and set it in kconfig.

Mostly it just requries some thought and design, I suspect.  I'm not the hardware model expert, really.
